### DISCUSSION TITLE
Custom modal dialog

### DISCUSSION THREAD

**Discussion by pomozoff** [pointfreeco/swift-composable-architecture#2580] [Q&A]
I want to present a custom modal view with a custom transition.
But I do not know how to wrap a custom SUI view modifier that presents a custom modal dialog with a store scoped to a substate, as same as the `sheet` modifier already wrapped in the TCA library?
Basically, I'm able to scope the substate from the store inside the content closure, however, I want to leverage Presentation property wrappers for a state and an action related to the substate.
Can't find anything related to the issue.
Any help appreciated!

**Comment by mbrandonw** [comment]
Hi @pomozoff, currently we have the presentation APIs that power our `sheet(store:)`, `popover(store:)` etc hidden behind `@_spi(Presentation)`. So, you can do an SPI import to get access to all of those tools:

```swift
@_spi(Presentation) import ComposableArchitecture
```

And you should be able to follow what we do in our navigation view modifiers to make your own.

The reason it is hidden behind SPI is because we weren't quite sure they were the right shape and so we wanted to understand people's use cases of the APIs before we fully publicized. However, since then we feel that they are probably just fine to use as-is, and further, with the [observation-beta](https://github.com/pointfreeco/swift-composable-architecture/discussions/2594), all of these tools are obsoleted. You can just use the regular `sheet(item:)`, `popover(item:)` etc APIs, and you can make custom navigation view modifiers in a way that is more similar to vanilla SwiftUI, and it will integrate into TCA immediately with no additional work.

**Reply by pomozoff** [reply]
Brandon, thank you so much! You are awesome, guys!

**Reply by KangKyung** [reply]
That really helped me a lot! You are so amazing people. I respect you🙇🏻‍♂️
